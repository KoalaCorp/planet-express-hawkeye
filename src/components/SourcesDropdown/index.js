import React, { useState, useEffect } from "react"

import PropTypes from "prop-types"

import DropdownError from "../DropdownError"
import DropdownMenu from "../DropdownMenu"
import DropdownButton from "../DropdownButton"
import Sources from "../Sources"

import Styled from "./styled"

import useFetchSources from "../../hooks/fetch-sources"

const SourcesDropdown = ({ initiallySelectedSources }) => {
  const { loadingSources, sources, errorsFetchSources } = useFetchSources()
  const [selectedSources, setSelectedSources] = useState(
    initiallySelectedSources
  )
  const [collapsed, setCollapsed] = useState(true)
  const [dropdownPrompt, setDropdownPrompt] = useState("all sources selected")

  useEffect(() => {
    if (!loadingSources) {
      const message =
        selectedSources.length === sources.length
          ? "all sources selected"
          : selectedSources.length > 0
          ? `${selectedSources.length} sources selected`
          : "No sources selected"
      setDropdownPrompt(message)
    }
  }, [loadingSources, selectedSources, sources])

  useEffect(() => {
    setSelectedSources(
      initiallySelectedSources.length
        ? initiallySelectedSources
        : sources.map((el) => el.attributes.label)
    )
  }, [initiallySelectedSources, sources])

  const handleSourceInputChange = (e) => {
    if (e.target.checked) {
      setSelectedSources([...selectedSources, e.target.value])
    } else {
      setSelectedSources(
        selectedSources.filter((source) => source !== e.target.value)
      )
    }
  }

  const handleClickButton = (e) => {
    e.preventDefault()
    setCollapsed(!collapsed)
  }

  const hasSelectedSources = () => {
    return selectedSources.length
  }

  return (
    <Styled.SourcesDropdown
      className="dropdown dropdown--sources"
      collapsed={collapsed}
      loadingSources={loadingSources}
      hasSelectedSources={hasSelectedSources()}
    >
      <DropdownButton onClick={handleClickButton}>
        {dropdownPrompt}
      </DropdownButton>

      <DropdownMenu>
        {!loadingSources && !hasSelectedSources() && (
          <DropdownError>
            You have to select at least one medium to search
          </DropdownError>
        )}

        <Sources
          sources={sources}
          selectedSources={selectedSources}
          errors={errorsFetchSources}
          handleSourceInputChange={handleSourceInputChange}
        />
      </DropdownMenu>
    </Styled.SourcesDropdown>
  )
}

export default SourcesDropdown

SourcesDropdown.defaultProps = {
  initiallySelectedSources: [],
}

SourcesDropdown.propTypes = {
  initiallySelectedSources: PropTypes.arrayOf(PropTypes.string),
}
