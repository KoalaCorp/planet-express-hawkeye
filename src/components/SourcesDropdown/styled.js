import styled from "styled-components"
import { rgba } from "polished"

const SourcesDropdown = styled.div`
  position: relative;

  &:not(.dropdown--collapsed) {
    .dropdown__button {
      background: ${(props) => rgba(props.theme.settings.colors.dark, 0.1)};
      color: ${(props) =>
        rgba(props.theme.settings.colors.palette.primary, 0.8)};
    }
  }

  &.dropdown--collapsed {
    .dropdown__menu {
      display: none;
    }

    .fa-caret-up {
      transform: translateY(-50%) rotate(180deg);
    }
  }
`

export default { SourcesDropdown }
