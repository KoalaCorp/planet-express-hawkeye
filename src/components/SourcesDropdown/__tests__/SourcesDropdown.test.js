import React from "react"
import { render } from "../../../../tests/helpers"

import SourcesDropdown from ".."

const renderComponent = ({ ...props } = {}) =>
  render(<SourcesDropdown {...props} />)

describe("SourcesDropdown", () => {
  it("renders correctly", () => {
    const dropdown = renderComponent()
    expect(dropdown).toMatchSnapshot()
  })
})
