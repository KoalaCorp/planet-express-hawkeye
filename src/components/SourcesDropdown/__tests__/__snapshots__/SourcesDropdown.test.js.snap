// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourcesDropdown renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="styled__Dropdown-sc-191offi-0 jQa-dxL dropdown dropdown--sources dropdown--collapsed"
      >
        <button
          class="styled__Button-sc-9epqn7-0 iWbztU dropdown__button"
        >
          all sources selected
          <i
            icon="[object Object]"
          />
        </button>
        <div
          class="styled__Menu-uutoje-0 bdYxwB dropdown__menu"
        >
          <div
            class="DropdownError__StyledDropdownError-sc-1pi6o8b-0 geVLwp dropdown__error dropdown__error--visible"
          >
            <i
              icon="[object Object]"
            />
            You have to select at least one medium to search
          </div>
          <div
            class="sources"
          >
            <div
              class="sources__errors"
            >
              TypeError: Cannot read property 'planetExpressAbe' of undefined
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="styled__Dropdown-sc-191offi-0 jQa-dxL dropdown dropdown--sources dropdown--collapsed"
    >
      <button
        class="styled__Button-sc-9epqn7-0 iWbztU dropdown__button"
      >
        all sources selected
        <i
          icon="[object Object]"
        />
      </button>
      <div
        class="styled__Menu-uutoje-0 bdYxwB dropdown__menu"
      >
        <div
          class="DropdownError__StyledDropdownError-sc-1pi6o8b-0 geVLwp dropdown__error dropdown__error--visible"
        >
          <i
            icon="[object Object]"
          />
          You have to select at least one medium to search
        </div>
        <div
          class="sources"
        >
          <div
            class="sources__errors"
          >
            TypeError: Cannot read property 'planetExpressAbe' of undefined
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
